src/policy.c

```diff
@@ -0,0 +1,421 @@
+        vm_matches_rule(rule, vm))
+      return (rule->cmd != DENY);
+  }
+
+  /* No match found, default to DENY. Return TRUE here to default to ALLOW */
+  return false;
+}
+
+/**
+ * This function should be called when a new device is plugged.
+ * It will assign the device to a VM according to policy.
+ *
+ * @param device A pointer to the device that was just plugged
+ *
+ * @return 1 if the device didn't get plugged to anything, the result
+ *         of usbowls_plug_device otherwise.
+ */
+int
+policy_auto_assign_new_device(device_t *device)
+{
+  rule_t *rule;
+  vm_t *vm = NULL;
+  int uivm;
+
+  /* If there's a sticky rule for the device, assign the the
+   * corresponding VM (if it's running). If there's no sticky rule
+   * for the device, consider assigning it to the focused VM */
+  rule = sticky_lookup(device);
+  if (rule != NULL) {
+    vm = vm_lookup_by_uuid(rule->vm_uuid);
+  } else {
+    if (vm == NULL) {
+      vm = vm_focused();
+      if (vm != NULL && !vm_gets_devices_when_in_focus(vm))
+        vm = NULL;
+    }
+  }
+
+  property_get_com_citrix_xenclient_xenmgr_vm_domid_(g_xcbus, XENMGR, UIVM_PATH, &uivm);
+  if (vm != NULL &&
+      vm->domid > 0 &&
+      vm->domid != uivm &&
+      policy_is_allowed(device, vm))
+  {
+    int res;
+
+    device->vm = vm;
+    res = usbowls_plug_device(vm->domid, device->busid, device->devid);
+    if (res != 0)
+      device->vm = NULL;
+    return res;
+  }
+
+  return 1;
+}
+
+/**
+ * Iterate over all the sticky rules that match the VM, and assign the
+ * corresponding devices to it
+ *
+ * @param vm The VM that just started
+ *
+ * @return 0 for success, -1 if anything went wrong
+ */
+int
+policy_auto_assign_devices_to_new_vm(vm_t *vm)
+{
+  struct list_head *pos, *device_pos;
+  rule_t *rule;
+  device_t *device;
+  int ret = 0;
+
+  /* For all the ALWAYS rules that match the VM, assign all devices
+   * that match the rule to the VM */
+  list_for_each(pos, &rules.list) {
+    rule = list_entry(pos, rule_t, list);
+    if (rule->cmd == ALWAYS && !strcmp(rule->vm_uuid, vm->uuid)) {
+      list_for_each(device_pos, &devices.list) {
+        device = list_entry(device_pos, device_t, list);
+        if (!device_matches_rule(rule, device))
+          continue;
+        /* The device matches the rule, let's try to assign it */
+        if (device->vm != NULL) {
+          if (device->vm != vm)
+          {
+            xd_log(LOG_ERR, "An always-assign device is assigned to another VM, this shouldn't happen!");
+            ret = -1;
+            continue;
+          } else {
+            /* The device is already assigned to the right VM */
+            continue;
+          }
+        } else {
+          /* The device is not assigned, as expected, plug it to its VM */
+          /* No need to check the policy, ALWAYS implies ALLOW */
+          device->vm = vm;
+          ret |= usbowls_plug_device(vm->domid, device->busid, device->devid);
```

Comment:
docs say usbowls_plug_device returns 1 on failure, yet this function is supposed to return -1. I'm not sure what's happening here, but I suspect something is amiss :)